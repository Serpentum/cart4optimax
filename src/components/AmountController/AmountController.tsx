import React, {ChangeEvent, CSSProperties, useEffect, useState} from 'react';
import s from './s.module.css'
import Button from "../UI/Button";

export interface AmountControllerI {
  value: number
  setter: (newValue?: number) => void
}

const customBtnJSS: CSSProperties = {
  padding: 0,
  width: 30,
  height: 30,
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
}

export const onChangeCurtCounters = (e: ChangeEvent<HTMLInputElement>, setter: (x: number) => void) => {
  const toNum = Number(e.currentTarget.value)
  setter(toNum)
}

const AmountController = ({value, setter}: AmountControllerI): JSX.Element => {

  const [localVal, setLocalVal] = useState<number>(value)

  const increase = () => {
    if(Number(value) + 1 < 100) {
      setter(Number(value) + 1)
    }
  }
  const decrease = () => {
    if(Number(value) - 1 > 0) {
      setter(Number(value) - 1)
    }
  }

  const onBlur = () => {
    if(localVal > 99) {
      setter(99)
      return
    }
    if(localVal <= 0) {
      setter(1)
      return
    }
    setter(localVal)
  }

  const onChange = (e: ChangeEvent<HTMLInputElement>) => onChangeCurtCounters(e, setter)

  useEffect(() => {
    setLocalVal(value)
  }, [value])

  return (
    <div className={s.container}>
      <Button.Default
        label={"+"}
        onClick={increase}
        customStyle={customBtnJSS}
      />
      <input
        className={s.input}
        type="number"
        value={localVal}
        onChange={onChange}
        onBlur={onBlur}
      />
      <Button.Default
        label={"-"}
        onClick={decrease}
        customStyle={customBtnJSS}
      />
    </div>
  );
};

export default AmountController;
