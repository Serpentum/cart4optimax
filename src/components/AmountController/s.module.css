.container{
    display: flex;
}

.input{
    width: 30px;
    border: none;
    text-align: center;
}

