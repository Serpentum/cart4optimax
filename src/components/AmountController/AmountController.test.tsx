import React from "react";
import {configure, shallow, mount} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import AmountController from "./index";

configure({ adapter: new Adapter() })

const setter = (value:number | undefined) => value
const initial = {
  value: 0,
  setter
}

describe('<AmountController />', () => {
  const wrapper = mount(<AmountController {...initial} />)

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('be rendered', () => {
    expect(wrapper.find('div>input[type="number"]').length).toBe(1)
  })

  it('should increase', () => {
    let value = 1
    const changeValue = jest.fn(val => {value = val});
    const wrapper = mount(<AmountController value={value} setter={changeValue} />);
    wrapper.find('div>#increaseBtn').simulate("click")
    expect(changeValue).toBeCalled();
    expect(value).toEqual(2)
  })

  it('should decrease', () => {
    let value = 2
    const changeValue = jest.fn(val => {value = val});
    const wrapper = mount(<AmountController value={value} setter={changeValue} />);
    wrapper.find('div>#decreaseBtn').simulate("click")
    expect(changeValue).toBeCalled();
    expect(value).toEqual(1)
  })

  it('should not start setter if val > 99', () => {
    let value = 99
    const changeValue = jest.fn(val => {value = val});
    const wrapper = mount(<AmountController value={value} setter={changeValue} />);
    wrapper.find('div>#increaseBtn').simulate("click")
    expect(changeValue).toBeCalledTimes(0);
    expect(value).toEqual(99)
  })

  it('should not start setter if val < 1', () => {
    let value = 1
    const changeValue = jest.fn(val => value= val);
    const wrapperMounted = mount(<AmountController value={value} setter={changeValue} />);
    wrapperMounted.find('div>#decreaseBtn').simulate("click")
    expect(changeValue).toBeCalledTimes(0);
    expect(value).toEqual(1)
  })
})

