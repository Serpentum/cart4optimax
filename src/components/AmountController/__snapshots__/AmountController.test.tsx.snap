// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountController /> should match the snapshot 1`] = `"<div class=\\"container\\"><button id=\\"increaseBtn\\" class=\\"container\\" type=\\"button\\" style=\\"padding: 0px; width: 30px; height: 30px; display: flex; justify-content: center; align-items: center;\\" tabindex=\\"0\\"><span class=\\"label\\">+</span></button><input class=\\"input\\" type=\\"number\\" value=\\"0\\"><button id=\\"decreaseBtn\\" class=\\"container\\" type=\\"button\\" style=\\"padding: 0px; width: 30px; height: 30px; display: flex; justify-content: center; align-items: center;\\" tabindex=\\"0\\"><span class=\\"label\\">-</span></button></div>"`;
